### PR TITLE
PAINTROID-736: Implement Shapes Tool

### PR DESCRIPTION
**Pull request by bakicelebi**
[PAINTROID-736](https://catrobat.atlassian.net/browse/PAINTROID-736)

## New Features and Enhancements
- Implemented BoundingBox class to control the transformation of the shape tool and in future text, clipboard and transform tools aswell
- Implement shapes tool provider that puts the bounding box in canvas center
- Implemented abstract ShapeCommand with SquareShapeCommand and CircleShapeCommand extending it.
  - Added both new commands to CommandFactory
  - Added fromJson cases for both new commands in Command class
  - Added serializer versions and types for both new commands
  - Added new checks in getNextTool in CommandManager for both new commands
- CommandPainter draws current shape on the canvas based on currentbounding box of shape tool
- Added static Paint objects to retrieve shape tool guide line paints
  - boundingBoxPaint
  - circumferencePaint
  - anchorPointPaint
- Added BoundingBoxCorner and ShapeType enums
- Extended Offset class to help with BoundingBox calculations
- Implemented isRotatingBoundingBoxProvider to lock 
BoundingBox only to rotation transformation
- Wrapped the clearRedoStack call in an if currentTool is shapes tool block because the shape command is finalized and added to command manager stack when checkmark is clicked
- Added Paint to onCheckmark method to be able to paint the shapes with current paint
- Added constants file to const variables such as pi divided by 2
- Added shape tool option widgets and shape tool option case in ToolOptions widget builder
  - ShapesToolShapeTypeOptions
  - ShapesToolTransformationModeOptions
- Changed ref.read(paintProvider) in canvas painter to ref.watch to reflect paint changes live in shape tool preview
- Added shapes tool onCheckmark logic in TopAppBar
- Added unit, widget and integration tests for all new features
## Refactorings and Bug Fixes
- Removed currentToolType from toolBoxStateProvider as it is a reduntant field. Current tool type can be retrieved from currentTool itself
- Changed toolBoxStateData builder to use ref.read instead of ref.watch because we do not want to invalidate the state whenever brushToolProvider is invalidated.
-Changed StrokeWidth slide inactive color to grey for visual clarity

#### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Paintroid/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.

- [x] Include the name of the Jira ticket in the PR’s title
- [x] Add the link to the ticket in Jira in the description of the PR
- [x] Include a summary of the changes plus the relevant context
- [x] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines (Wiki)
- [x] Verify that the changes generate no compiler or linter warnings
- [x] Perform a self-review of the changes
- [x] Verify to commit no other files than the intentionally changed ones
- [x] Include reasonable and readable tests verifying the added or changed behavior
- [x] Confirm that new and existing tests pass locally
- [x] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [x] Verify that your changes do not have any conflicts with the base branch
- [x] After the PR, verify that all CI checks have passed
- [x] Add new information to the [Wiki](https://github.com/Catrobat/Paintroid-Flutter/wiki)

[PAINTROID-736]: https://catrobat.atlassian.net/browse/PAINTROID-736?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: test/utils/widget_finder.dart [hunk 1]
```diff
@@ -21,4 +21,7 @@ class WidgetFinder {
       find.byKey(ValueKey(TopBarActionData.UNDO.name));
   static final Finder redoButton =
       find.byKey(ValueKey(TopBarActionData.REDO.name));
+
+  static final Finder circleShapeTypeChip =
+      find.byKey(const ValueKey('circle_shape_type_chip'));
```
Comment (Lenkomotive) [inline test/utils/widget_finder.dart]: pls add the key to the `WidgetIdentifier` file and use the const value

**Review by Lenkomotive — CHANGES_REQUESTED** [review]
Great work! 

A few things that i noticed while using the app:

1. the width of a shape shrinks with increasing strokeWidth
2. rotating around the anchor while in scale mode makes it hard to position the shape correctly horizontal.
Does it make sense to allow rotation only in when rotation mode is on? And otherwise just scale.
3. When moving the shape, it would be good maybe if possible that the center of the actual shape is not changed. This way the user can click outside the shape to move it without it jumping to the newly clicked position.

If anything is unclear pls feel free to text me!

**Comment by bakicelebi** [comment]
> 2\. rotating around the anchor while in scale mode makes it hard to position the shape correctly horizontal.
>         Does it make sense to allow rotation only in when rotation mode is on? And otherwise just scale.

This is going to require a lot more changes and I think the the ticket is already to big and complex to review. We should create a separate ticket for this.

**Review by Lenkomotive — APPROVED** [review]

File: lib/core/utils/constants.dart [hunk 2]
```diff
@@ -0,0 +1 @@
+const double HALF_PI = 1.5707963267948966;
```
Comment (juliajulie95) [inline lib/core/utils/constants.dart]: Würd ich PI / 2 schreiben, also denk PI kann man sicher importen von math oder?

File: lib/ui/pages/workspace_page/components/bottom_bar/tool_options/shapes_tool/shapes_tool_shape_type_options.dart [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+                                ? PaintroidTheme.of(context).primaryColor
+                                : Colors.white,
+                        chipIcon: Icon(
+                          Icons.circle_outlined,
+                          color: PaintroidTheme.of(context).shadowColor,
+                        ),
+                        onPressed: () {
+                          shapesTool.shapeType = ShapeType.circle;
+                          ref.read(canvasPainterProvider.notifier).repaint();
+                          setState(() {});
```
Comment (juliajulie95) [inline lib/ui/pages/workspace_page/components/bottom_bar/tool_options/shapes_tool/shapes_tool_shape_type_options.dart]: Wollen wir statt dem set state noch einen Provider machen?
Ich mein von der performance her wärs okay, weils eh das ganze Widget betrifft, aber weiß net

**Comment by juliajulie95** [comment]
Good work, thank you!

**Pull request merged**